### PR TITLE
Genetic algorithm optimizers

### PR DESCRIPTION
**Pull request by jmgoff**
There has been a need and requests (e.g. through Issue # 35 Parameter Optimization) to include some sort of native optimizer for FitSNAP fits. This pull request addresses this as well as some other minor corrections to new examples:

1. **a)** Optimization algorithms (genetic algorithm, simmulated annealing) have been implemented and optimize hyperparameters such as the group weights by using descriptor design matrices stored in memory (enabled by the library mode) which significantly reduces computational cost and file I/O compared to other optimization methods. These procedures optimize hyperparameters with respect to a cost function, which by default is RMSE energy + RMSE force.  **b)** Examples of hyperparameter optimization are given for both SNAP and ACE models of Ta. **c)** Functionality has been added to these optimization procedures so that users may add simple objective functions to the default cost function, such as energy differences between two structures. This is useful for users who want to fit directly to a property such as adsorption energy. 

2. Bug-fixes and other corrections suggested by Meg McCarthy have been included (fixing paths to data files for new ACE examples and tests for GA examples). New ACE examples and GA examples should run from machine to machine.

3. Script to generate _ball-park_ defaults on hyperparameters for: ACE models, ZBL reference potentials, and more based on physical properties such as ionic bond lengths, has been added to the 'tools' directory.  

4. Documentation for FitSNAP + lammps_compute_PACE has been updated. Some users have reported issues with certain choices of D flags during lammps compilation - these issues don't arise until FitSNAP calls LAMMPS. An example of alternative D flags has been provided that work for cases where the defaults fail.

The optimization algorithms implemented run in serial, but the processing of descriptors may still be done in parallel.

### CODE REVIEW

File: examples/library/genetic_algorithm_ACE/24Apr23_Standard/Ta.in [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+detailed_errors = 1
+
+[RIDGE]
+alpha = 1.E-5
+
+[SCRAPER]
+scraper = JSON
+
+[PATH]
+dataPath = /home/jmgoff/work/new_Ta_ex/single_elm_grid/JSON
```
Comment (jan-janssen) [inline examples/library/genetic_algorithm_ACE/24Apr23_Standard/Ta.in:37]: This path seems to be wrong

Comment (jan-janssen) [reply examples/library/genetic_algorithm_ACE/24Apr23_Standard/Ta.in:37]: suggestion```
dataPath = ../../Ta_Linear_JCP2014/JSON
```

File: examples/library/genetic_algorithm_SNAP/libmod_optimize.py [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+            #NOTE to add another contribution to the cost function , you need to evaluate it in the loop
+            # and add it to the fit_and_cost function
+            # if this involves a lammps simulation, you will have to print potentials at the different steps
+            # to run the lammps/pylammps simulation. To do so, the fitsnap output file name prefix should
+            # be updated per step, then snap.write_output() should be called per step. This will likely increase
+            # the optimization time.
+        # Anything printed with pt.single_print will be included in output file.
+        pt.single_print('generation,scores,popsize:',generation,len(scores),population_size)
+        # Print generation and best fit.
+        bestfit = min(scores)
```
Comment (rohskopf) [inline examples/library/genetic_algorithm_SNAP/libmod_optimize.py:510]: Added this as a suggestion - is this the best way to track progress?
There's a lot of screen output so I dumped that into a file with the `screen2file` command line option.
We can use normal `print` to only output important info to the screen.

**Review by mitwood — APPROVED** [review]
Provides the addition of optimization tools for fitting models, utilizes library calls to functions of the core code to prove a streamlined workflow to achieve this. Designed for experiences users.

My main cause for concern is that the optimization tools are not included in the core code, rather they are isolated to what is effectively an external script. If the gc and simulated annealing functions can be moved to an existing class (solver?) or a new one, that are then called by the users' driver script, that would be ideal. The examples have large chunks of duplicated code that can be moved in the future, likely after #213 is completed.  I dont mind holding onto these changes as examples for now before they are moved, but we need to be careful not to balloon the examples folder with too much content.

**Pull request merged**